src/sparseml/exporters/transforms/kv_cache/configs.py

```diff
@@ -223,10 +223,51 @@ def get_kv_cache_config(
     kv_cache_config.num_attention_heads = num_attention_heads
     kv_cache_config.hidden_size_kv_cache = hidden_size_kv_cache
 
+    kv_cache_config = adapt_cache_structure_for_gqa(
+        kv_cache_config, transformers_config
+    )
+
     _LOGGER.info("Properly configured arguments for KV Cache Transformation")
     return kv_cache_config
 
 
+def adapt_cache_structure_for_gqa(
+    kv_cache_config: KeyValueCacheConfig,
+    transformers_config: Dict[str, Any],
+    model_names: List[str] = ["llama"],
+) -> KeyValueCacheConfig:
+    """
+    Potentially adapts the kv_cache_config,so that it
+    properly works with Grouped Query Attention (GQA).
+
+    :param kv_cache_config: The kv cache config for the model.
+    :param transformers_config: The transformers config for
+        the model. If contains the key:`num_key_value_heads`,
+        the model may be potentially using GQA instead of
+        Multi Query Attention (MQA) and thus the kv_cache_config
+        needs to be adapted.
+    :param model_names: The list of model names that may use
+        GQA instead of MQA.
+    :return: Potentially adapted kv cache config for the model.
+        If the model does not use GQA, the kv_cache_config is
+        returned unchanged.
+    """
+    num_key_value_heads = transformers_config.get("num_key_value_heads")
+
+    if num_key_value_heads is not None:
+        # If num_key_value_heads=1 the model will use
+        # Multi Query Attention (MQA) otherwise GQA is used.
+        if kv_cache_config.model_name in model_names and num_key_value_heads > 1:
+            # For now, we only support GQA for LLAMA.
+            kv_cache_config.transpose_value_input = None
+
+            _LOGGER.info(
+                f"Adapted the model: {transformers_config['model_type']} "
+                f"to work with GQA."
+            )
```

Comment:
there is also the case where `num_key_value_heads==num_attention_heads`, where we are using the original multi-head attention

> If `num_key_value_heads=num_attention_heads`, the model will use Multi Head Attention (MHA), if `num_key_value_heads=1` the model will use Multi Query Attention (MQA) otherwise GQA is used.

see the config.json for llama7b
```
"num_attention_heads": 32,
"num_key_value_heads": 32,
```

Reply:
@mgoin I got you.